src/energyhub.py

```diff
@@ -735,136 +754,247 @@ def _write_solution_diagnostics(self, save_path):
-                unit_capex * annualization_factor
-            )
-
-            # Remove constraint (from persistent solver and from model)
-            self.solver.remove_constraint(b_tec.const_capex_aux)
-            b_tec.del_component(b_tec.const_capex_aux)
+            if MC_technology_row is not None:
+                unit_capex = random.uniform(
+                    MC_technology_row["min"], MC_technology_row["max"]
+                )
+            else:
+                unit_capex = tec_data.economics.capex_data["unit_capex"] * sd_random
 
-            # Add constraint again
-            b_tec.const_capex_aux = Constraint(
-                expr=b_tec.var_size * b_tec.para_unit_capex_annual
-                == b_tec.var_capex_aux
-            )
-            self.solver.add_constraint(b_tec.const_capex_aux)
+            b_tec.para_unit_capex = unit_capex
+            b_tec.para_unit_capex_annual = unit_capex * annualization_factor
 
-        elif capex_model == 2:
+        elif tec_data.economics.capex_model == 2:
             warnings.warn(
                 "monte carlo on piecewise defined investment costs is not implemented"
             )
 
-        elif capex_model == 3:
-            # TODO capex model 3
-            warnings.warn(
-                "monte carlo on piecewise defined investment costs is not implemented"
-            )
+        elif tec_data.economics.capex_model == 3:
+            if MC_technology_row is not None:
+                for row in MC_technology_row:
+                    if row["parameter"] == "unit_CAPEX":
+                        unit_capex = random.uniform(row["min"], row["max"])
+                    elif row["parameter"] == "fix_CAPEX":
+                        fix_capex = random.uniform(row["min"], row["max"])
+            else:
+                unit_capex = tec_data.economics.capex_data["unit_capex"] * sd_random
+                fix_capex = tec_data.economics.capex_data["fix_capex"] * sd_random
+
+            b_tec.para_unit_capex = unit_capex
+            b_tec.para_unit_capex_annual = unit_capex * annualization_factor
+            b_tec.para_fix_capex = fix_capex
+            b_tec.para_fix_capex_annual = fix_capex * annualization_factor
+
+        # Change variable bounds
+        def calculate_max_capex():
+            if economics.capex_model == 1:
+                max_capex = b_tec.para_size_max * b_tec.para_unit_capex_annual
+                bounds = (0, max_capex)
+            elif economics.capex_model == 2:
+                # Todo: Calculate
+                pass
+            elif economics.capex_model == 3:
+                max_capex = (
+                    b_tec.para_size_max * b_tec.para_unit_capex_annual
+                    + b_tec.para_fix_capex_annual
+                )
+                bounds = (0, max_capex)
+            else:
+                bounds = None
+            return bounds
+
+        bounds = calculate_max_capex()
```

Comment:
Wait, this we have double now, or not? We also have it in the technology.py?

Reply:
Same as previous comment, do you want to have it like the networks or like the technologies?